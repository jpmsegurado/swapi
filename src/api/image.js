import axios from 'axios'

const url = `https://www.googleapis.com/customsearch/v1?cx=${process.env.VUE_APP_GOOGLE_API_CX}&q=luke%20skywalker&key=${process.env.VUE_APP_GOOGLE_API_KEY}&searchType=image`

const Image = {
  getFromGoogle (search) {
    return axios.get(url, { params: { q: search } })
  }
}

export default Image
