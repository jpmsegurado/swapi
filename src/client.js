import axios from 'axios'

const client = axios.create({
  baseURL: 'https://swapi.co/api/',
  timeout: 10000
})

export default client
