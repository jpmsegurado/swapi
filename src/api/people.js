import client from '../client'
import mock from './mock'
import Image from './image'
const json = localStorage.getItem('people')
const cache = json ? JSON.parse(json) : {}

const People = {
  async getAll ({ page = 1, search }) {
    const params = { page }
    if (search) { params.search = search }

    if (cache[JSON.stringify(params)]) { return Promise.resolve(cache[JSON.stringify(params)]) }

    try {
      const response = await client.get('/people', { params })
      let items = response.data.results
      items = await Promise.all(items.map(async (item) => {
        try {
          const { data } = await Image.getFromGoogle(item.name.toLowerCase())
          item.image = data.items[0].link
        } catch (e) {
          item.image = 'https://upload.wikimedia.org/wikipedia/commons/6/6d/Star_Wars_Logo_2.svg'
        }

        const id = parseInt(item.url.match(/[0-9]{1,5}/))
        item.id = id
        return item
      }))

      response.data.results = items
      cache[JSON.stringify(params)] = response.data
      localStorage.setItem('people', JSON.stringify(cache))
      return response.data
    } catch (e) {
      return mock
    }
  },
  async get (id) {
    const json = localStorage.getItem(`person/${id}`)
    if (json) { return JSON.parse(json) }

    const response = await client.get(`/people/${id}`)
    const item = response.data

    try {
      const { data } = await Image.getFromGoogle(item.name.toLowerCase())
      item.image = data.items[0].link
    } catch (e) {
      item.image = 'https://upload.wikimedia.org/wikipedia/commons/6/6d/Star_Wars_Logo_2.svg'
    }

    item.id = parseInt(id)

    localStorage.setItem(`person/${id}`, JSON.stringify(item))

    return item
  }
}

export default People
