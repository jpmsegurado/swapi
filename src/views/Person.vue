<template>
  <div
    v-loading="loading"
    element-loading-text="Loading..."
    element-loading-spinner="el-icon-loading"
    element-loading-background="rgba(0, 0, 0, 0.8)"
    class="person"
  >
    <div class="person__profile">
      <div class="person__profile__image">
        <img :src="person.image">
      </div>

      <h2 class="person__profile__name">
        {{ person.name }}
      </h2>

      <p class="person__profile__field">
        Birth Year: <strong>{{ person.birth_year }}</strong>
      </p>

      <p class="person__profile__field">
        Eye color: <strong>{{ person.eye_color }}</strong>
      </p>

      <p class="person__profile__field">
        Gender: <strong>{{ person.gender }}</strong>
      </p>

      <p class="person__profile__field">
        Hair color: <strong>{{ person.hair_color }}</strong>
      </p>

      <p class="person__profile__field">
        Height: <strong>{{ person.height }}cm</strong>
      </p>

      <p class="person__profile__field">
        Mass: <strong>{{ person.mass }}kg</strong>
      </p>

      <p class="person__profile__field">
        Skin color: <strong>{{ person.skin_color }}</strong>
      </p>
    </div>

    <div v-if="!loading" class="person__content">
      <div class="person__content__section">
        <h2>Species</h2>
        <p v-for="specie in person.species" :key="specie.url">
          {{ species }}
        </p>
      </div>

      <div class="person__content__section">
        <h2>Homeworld</h2>
        <p>{{ person.homeworld.name }}</p>
      </div>

      <div class="person__content__section">
        <h2>Films</h2>
        <p>
          {{ films }}
        </p>
      </div>

      <div class="person__content__section">
        <h2>Vehicles</h2>
        <p>
          {{ vehicles.length > 2 ? vehicles : 'No Vehicles' }}
        </p>
      </div>

      <div class="person__content__section">
        <h2>Starship</h2>
        <p>
          {{ starships.length > 2 ? starships : 'No Starships' }}
        </p>
      </div>
    </div>
  </div>
</template>

<script>
import client from '../client'
import People from '@/api/people'

export default {
  data () {
    return {
      loading: true,
      person: {}
    }
  },
  computed: {
    films () {
      return this.person.films.map(film => film.title).join(', ')
    },
    species () {
      return this.person.species.map(specie => specie.name).join(', ')
    },
    vehicles () {
      return this.person.vehicles.map(vehicle => vehicle.name).join(', ')
    },
    starships () {
      return this.person.starships.map(starship => starship.name).join(', ')
    }
  },
  async mounted () {
    this.person = await People.get(this.$route.params.id)
    await this.loadWorld(this.person.homeworld)
    await this.loadFilms(this.person.films)
    await this.loadSpecies(this.person.species)
    await this.loadStarships(this.person.starships)
    this.loading = false
  },
  methods: {
    async loadWorld (url) {
      const response = await client.get(url)
      this.person.homeworld = response.data
    },
    async loadFilms (filmsUrl) {
      if (filmsUrl.length === 0) { this.person.films = [] }

      const films = await Promise.all(filmsUrl.map(async (url) => {
        const response = await client.get(url)
        const film = response.data
        return film
      }))

      this.person.films = films
    },
    async loadSpecies (speciesUrl) {
      if (speciesUrl.length === 0) { this.person.species = [] }

      const species = await Promise.all(speciesUrl.map(async (url) => {
        const response = await client.get(url)
        const specie = response.data
        return specie
      }))

      this.person.species = species
    },
    async loadVehicles (vehiclesUrl) {
      if (vehiclesUrl.length === 0) { this.person.vehicles = [] }

      const vehicles = await Promise.all(vehiclesUrl.map(async (url) => {
        const response = await client.get(url)
        const vehicle = response.data
        return vehicle
      }))

      this.person.vehicles = vehicles
    },
    async loadStarships (starshipsUrl) {
      if (starshipsUrl.length === 0) { this.person.starships = [] }

      const starships = await Promise.all(starshipsUrl.map(async (url) => {
        const response = await client.get(url)
        const starship = response.data
        return starship
      }))

      this.person.starships = starships
    }
  }
}
</script>

<style lang="scss" scoped>
.person {
  display: flex;
  position: absolute;
  width: 100%;
  height: 100%;
  padding: 20px 0;

  &__profile {
    width: 300px;
    padding: 20px;
    box-sizing: border-box;
    border-right: solid 1px rgba(255, 255, 255, 0.1);

    &__field {
      margin-top: 10px;
    }

    &__image {
      overflow: hidden;
      width: 250px;
      height: 250px;
      border-radius: 50%;
      margin: 0 auto;

      img {
        width: 100%;
        height: 100%;
        object-fit: cover;
      }
    }

    &__name {
      margin: 20px 0;
    }
  }

  &__content {
    width: calc(100% - 300px);
    padding: 20px;

    &__section {
      margin-bottom: 20px;

      h2 {
        margin-bottom: 10px;
      }

      p {
        margin-top: 10px;
        line-height: 25px;
      }
    }
  }
}

@media (max-width: $screen-sm) {
  .person {
    flex-wrap: wrap;

    &__profile {
      width: 100%;
      border: 0;
    }

    &__content {
      width: 100%;
    }
  }
}
</style>
