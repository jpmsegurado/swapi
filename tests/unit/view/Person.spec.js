import { shallowMount } from '@vue/test-utils'
import { person } from '@/api/mock'
import Person from '@/views/Person'
import client from '@/client'
import People from '@/api/people'

const $route = {
  params: {
    id: 14
  }
}

localStorage.getItem = jest.fn().mockImplementation(() => false)
localStorage.setItem = jest.fn().mockImplementation(() => false)

function setup () {
  localStorage.getItem = jest.fn().mockImplementation(() => false)
  localStorage.setItem = jest.fn().mockImplementation(() => false)
  return shallowMount(Person, {
    mocks: {
      $route
    },
    directives: {
      loading () {
        return true
      }
    }
  })
}

describe('Person', () => {
  let wrapper

  describe('#render', () => {
    it('renders correcly', () => {
      const wrapper = setup()
      expect(wrapper.element).toMatchSnapshot()
    })
  })

  describe('#computed', () => {
    beforeEach(() => {
      wrapper = setup()
    })

    describe('#films', () => {
      it('return two films titles separated by commas', () => {
        wrapper.setData({ person })
        expect(wrapper.vm.films).toBe('A New Hope, Rise of skywalker')
      })
    })

    describe('#species', () => {
      it('return two species separated by commas', () => {
        wrapper.setData({ person })
        expect(wrapper.vm.species).toBe('Robot, Droid')
      })
    })

    describe('#vehicles', () => {
      it('return two vehicles separated by commas', () => {
        wrapper.setData({ person })
        expect(wrapper.vm.vehicles).toBe('Falcon, X-wing')
      })
    })

    describe('#starships', () => {
      it('return two starships separated by commas', () => {
        wrapper.setData({ person })
        expect(wrapper.vm.starships).toBe('Falcon, X-wing')
      })
    })
  })

  describe('#methods', () => {
    beforeEach(() => {
      wrapper = setup()
    })

    describe('#loadWorld', () => {
      it('when called it calls client.get', () => {
        const spy = jest.spyOn(client, 'get')
        wrapper.vm.loadWorld('swapi.com')
        expect(spy).toHaveBeenCalled()
      })

      it('when called it sets person homeworld', async () => {
        const response = { data: { name: 'tatooine' } }
        client.get = jest.fn().mockResolvedValue(response)
        await wrapper.vm.loadWorld('swapi.com')
        expect(wrapper.vm.person.homeworld).toEqual({ name: 'tatooine' })
      })
    })

    describe('#loadFilms', () => {
      it('when called it sets person films', async () => {
        await wrapper.vm.loadFilms([])
        const films = []
        expect(wrapper.vm.person.films).toEqual(films)
      })

      it('when called with no urls it sets person films to empty array', async () => {
        const response = { data: { name: 'film 1' } }
        client.get = jest.fn().mockResolvedValue(response)
        await wrapper.vm.loadFilms(['swapi.com', 'swapi.com'])
        const films = [response.data, response.data]
        expect(wrapper.vm.person.films).toEqual(films)
      })
    })

    describe('#loadSpecies', () => {
      it('when called it sets person species', async () => {
        await wrapper.vm.loadSpecies([])
        const species = []
        expect(wrapper.vm.person.species).toEqual(species)
      })

      it('when called with no urls it sets person species to empty array', async () => {
        const response = { data: { name: 'specie 1' } }
        client.get = jest.fn().mockResolvedValue(response)
        await wrapper.vm.loadSpecies(['swapi.com', 'swapi.com'])
        const species = [response.data, response.data]
        expect(wrapper.vm.person.species).toEqual(species)
      })
    })

    describe('#loadVehicles', () => {
      it('when called it sets person vehicles', async () => {
        await wrapper.vm.loadVehicles([])
        const vehicles = []
        expect(wrapper.vm.person.vehicles).toEqual(vehicles)
      })

      it('when called with no urls it sets person vehicles to empty array', async () => {
        const response = { data: { name: 'vehicle 1' } }
        client.get = jest.fn().mockResolvedValue(response)
        await wrapper.vm.loadVehicles(['swapi.com', 'swapi.com'])
        const vehicles = [response.data, response.data]
        expect(wrapper.vm.person.vehicles).toEqual(vehicles)
      })
    })

    describe('#loadStarships', () => {
      it('when called it sets person starships', async () => {
        await wrapper.vm.loadStarships([])
        const starships = []
        expect(wrapper.vm.person.starships).toEqual(starships)
      })

      it('when called with no urls it sets person starships to empty array', async () => {
        const response = { data: { name: 'starship 1' } }
        client.get = jest.fn().mockResolvedValue(response)
        await wrapper.vm.loadStarships(['swapi.com', 'swapi.com'])
        const starships = [response.data, response.data]
        expect(wrapper.vm.person.starships).toEqual(starships)
      })
    })
  })

  describe('#mounted', () => {
    beforeEach(() => {
      wrapper = setup()
    })

    it('when called it sets person', async () => {
      wrapper.setMethods({
        loadWorld: jest.fn().mockResolvedValue({}),
        loadFilms: jest.fn().mockResolvedValue({}),
        loadSpecies: jest.fn().mockResolvedValue({}),
        loadStarships: jest.fn().mockResolvedValue({})
      })
      People.get = jest.fn().mockResolvedValue(person)
      await Person.mounted.call(wrapper.vm)
      expect(wrapper.vm.person).toBe(person)
    })
  })
})
