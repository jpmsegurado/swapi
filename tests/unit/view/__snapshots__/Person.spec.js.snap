// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Person #render renders correcly 1`] = `
<div
  class="person"
  element-loading-background="rgba(0, 0, 0, 0.8)"
  element-loading-spinner="el-icon-loading"
  element-loading-text="Loading..."
>
  <div
    class="person__profile"
  >
    <div
      class="person__profile__image"
    >
      <img />
    </div>
     
    <h2
      class="person__profile__name"
    >
      
      
    
    </h2>
     
    <p
      class="person__profile__field"
    >
      
      Birth Year: 
      <strong>
        
      </strong>
    </p>
     
    <p
      class="person__profile__field"
    >
      
      Eye color: 
      <strong>
        
      </strong>
    </p>
     
    <p
      class="person__profile__field"
    >
      
      Gender: 
      <strong>
        
      </strong>
    </p>
     
    <p
      class="person__profile__field"
    >
      
      Hair color: 
      <strong>
        
      </strong>
    </p>
     
    <p
      class="person__profile__field"
    >
      
      Height: 
      <strong>
        cm
      </strong>
    </p>
     
    <p
      class="person__profile__field"
    >
      
      Mass: 
      <strong>
        kg
      </strong>
    </p>
     
    <p
      class="person__profile__field"
    >
      
      Skin color: 
      <strong>
        
      </strong>
    </p>
  </div>
   
  <!---->
</div>
`;
